daemon/cmd/state.go

```diff
@@ -439,43 +444,52 @@ func (d *Daemon) allocateIPsLocked(ep *endpoint.Endpoint) (err error) {
 func (d *Daemon) initRestore(restoredEndpoints *endpointRestoreState) chan struct{} {
 	bootstrapStats.restore.Start()
 	var restoreComplete chan struct{}
+	var clusterSyncComplete chan struct{}
 	if option.Config.RestoreState {
 		// When we regenerate restored endpoints, it is guaranteed tha we have
 		// received the full list of policies present at the time the daemon
-		// is bootstrapped.
+		// is bootstrapped. Add a OneShot job to empty this channel
 		restoreComplete = d.regenerateRestoredEndpoints(restoredEndpoints)
-		go func() {
-			<-restoreComplete
-		}()
 
-		go func() {
-			if d.clientset.IsEnabled() {
-				// Also wait for all cluster mesh to be synchronized with the
-				// datapath before proceeding.
-				if d.clustermesh != nil {
-					err := d.clustermesh.ClustersSynced(d.ctx)
-					if err != nil {
-						log.WithError(err).Fatal("timeout while waiting for all clusters to be locally synchronized")
+		d.jobGroup.Add(
+			job.OneShot("empty restoreComplete", func(_ context.Context) error {
+				<-restoreComplete
+				return nil
+			}),
+			job.OneShot("setup-sync-lb-maps", func(ctx context.Context) error {
+				if d.clientset.IsEnabled() {
+					clusterSyncComplete = make(chan struct{}, 0)
+					// Also wait for all cluster mesh to be synchronized with the
+					// datapath before proceeding.
+					if d.clustermesh != nil {
+						err := d.clustermesh.ClustersSynced(d.ctx)
+						if err != nil {
+							log.WithError(err).Fatal("timeout while waiting for all clusters to be locally synchronized")
+						}
+						log.Debug("all clusters have been correctly synchronized locally")
 					}
-					log.Debug("all clusters have been correctly synchronized locally")
+					close(clusterSyncComplete)
 				}
-				// Start controller which removes any leftover Kubernetes
-				// services that may have been deleted while Cilium was not
-				// running. Once this controller succeeds, because it has no
-				// RunInterval specified, it will not run again unless updated
-				// elsewhere. This means that if, for instance, a user manually
-				// adds a service via the CLI into the BPF maps, that it will
-				// not be cleaned up by the daemon until it restarts.
-				controller.NewManager().UpdateController("sync-lb-maps-with-k8s-services",
-					controller.ControllerParams{
-						DoFunc: func(ctx context.Context) error {
-							return d.svc.SyncWithK8sFinished(d.k8sWatcher.K8sSvcCache.EnsureService)
-						},
-						Context: d.ctx,
-					},
-				)
-			}
-		}()
+				return nil
+			}),
+			// Start a job which removes any leftover Kubernetes
+			// services that may have been deleted while Cilium was not
+			// running. Since this is a OneShot job, it will not run again unless
+			// updated elsewhere. This means that if, for instance, a user manually
+			// adds a service via the CLI into the BPF maps, that it will
+			// not be cleaned up by the daemon until it restarts.
+			job.OneShot(
```

Comment:
Should we just avoid registering the job if the k8s client is disabled?

Reply:
makes sense, wrapped the job creation with clientsent Enabled check